feat: add support for describe statement

olavloite opened the pull request:
Adds support for DescribeStatement by transforming sql statements into a select statement that queries the (types of) the parameters in the original query. This can be used to temporarily support prepared statements until a DescribeStatement endpoint is available in the backend.

With this change the entire extended query protocol is supported by PGAdapter.

olavloite committed:
feat: add support for describe statement

Adds support for DescribeStatement by transforming sql statements into a
select statement that queries the (types of) the parameters in the
original query. This can be used to temporarily support prepared
statements until a DescribeStatement endpoint is available in the
backend.

olavloite committed:
feat: improve DescribeStatement support

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
test: add metadata results

olavloite committed:
test: fix pgx tests

olavloite committed:
perf: first describe portal

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
feat: execute actual select first

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
deps: use client lib SNAPSHOT

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
fix: remove local absolute path

olavloite committed:
fix: test failures

olavloite committed:
fix: JDBC metadata test

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

pratickchokhani commented on src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediatePreparedStatement.java:
```diff
@@ -159,34 +176,283 @@ public IntermediatePortalStatement bind(
+    if (parser.eat("select")) {
+      // This is an `insert into <table> [(...)] select ...` statement. Then we can just use the
+      // select statement as the result.
+      return Statement.of(
```
Can we make the Select statement method reusable?

olavloite replied:
Done

pratickchokhani commented on src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediatePreparedStatement.java:
```diff
@@ -159,34 +176,283 @@ public IntermediatePortalStatement bind(
+    }
+  }
+
+  private Statement transformSelectToSelectParams(Set<String> parameters) {
```
Can we add examples as comment? Same for other parsers

olavloite replied:
Done

pratickchokhani commented on src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediatePreparedStatement.java:
```diff
@@ -159,34 +176,283 @@ public IntermediatePortalStatement bind(
+  }
+
+  @VisibleForTesting
+  Statement transformToSelectParams(Set<String> parameters) {
```
Can this be made more simple by just fetching columns name and the respective table name? Because from what I understand the resultant query can be made from just those?
If yes, we can use regex to fetch the respective values. Which will be more readable and be less error prone.

olavloite replied:
No, that is not possible, as the type of a parameter might be different from the type of the column that it is used to for example filter. Consider for example the following statement:

```sql
select *
from my_table
where created_at > to_timestamp($1, 'YYYY-MM-DD HH:MI')
```

The type of the column `created_at` is `TIMESTAMP`, but the type of the parameter is `text`.

olavloite committed:
fix: address review comments

olavloite committed:
Merge branch 'postgresql-dialect' into test-describe-statement

olavloite committed:
fix: remove parameterized offset clause

Spangres does not (yet) support offset clauses that are not an integer literal.

olavloite committed:
Merge branch 'test-describe-statement' of github.com:GoogleCloudPlatform/pgadapter into test-describe-statement

olavloite committed:
fix: ignore test for unsupported backend feature
